File: programmingModels/OpenMP/demo/fortran/03_map_unstructured_function.f90
subroutine daxpy( a, b, scalar, num_elements )
  implicit none
  integer num_elements, j
  double precision :: a(num_elements), b(num_elements)
  double precision scalar

!$omp target teams distribute parallel do simd map(tofrom:a(:num_elements)) map(to:b(:num_elements))
    do j=1,num_elements
       a(j) = a(j)+scalar*b(j)
    end do
!$omp end target teams distribute parallel do simd

end subroutine daxpy

program main
  implicit none
  double precision  scalar
  integer err, j
  integer num_errors
  integer num_elements

  double precision, allocatable :: a(:)
  double precision, allocatable :: b(:)
  double precision, allocatable :: c(:)

  scalar = 8d0
  num_errors = 0
  num_elements = 1024

  allocate (a(num_elements),stat=err)
  allocate (b(num_elements),stat=err)
  allocate (c(num_elements),stat=err)

  ! initialize on the host
  do j=1,num_elements
     a(j) = 0d0
     b(j) = j
     c(j) = 0d0
  end do

!$omp target enter data map(to:a)
!$omp target enter data map(to:b)
!$omp target enter data map(to:c)

  call daxpy( a, b, scalar, num_elements )

  call daxpy( c, a, scalar, num_elements )

!$omp target update from(c)

  ! error checking
  do j=1,num_elements
     if( abs(c(j) - j*scalar*scalar) .gt. 0.000001 ) then
        num_errors = num_errors + 1
     end if

  end do

!$omp target exit data map(release:a)
!$omp target exit data map(release:b)
!$omp target exit data map(release:c)

  deallocate(a);
  deallocate(b);
  deallocate(c);

  if(num_errors == 0) then
    write(*,*) "Success!\n"
  else
    write(*,*) "Wrong!\n"
  endif

end program main

File: programmingModels/OpenMP/demo/cpp/03_map_unstructured_function.cpp
#include <math.h>
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

void daxpy( double * __restrict__ a, double * __restrict__ b,
	    double scalar, int num_elements ) {

#pragma omp target teams distribute parallel for map(tofrom:a[0:num_elements]) map(to:b[0:num_elements])
      for (size_t j=0; j<num_elements; j++) {
	a[j] = a[j] + b[j] * scalar;
      }

      return;
}

int main( int argc, char** argv )
{
  double*   a = NULL;
  double*   b = NULL;
  double*   c = NULL;
  double scalar = 8.0;
  int num_errors = 0;
  int num_elements = 1024;
  
  a = (double *) malloc( sizeof(double)*num_elements );
  b = (double *) malloc( sizeof(double)*num_elements );
  c = (double *) malloc( sizeof(double)*num_elements );

  // initialize on the host
  for (size_t j=0; j<num_elements; j++) {
      a[j] = 0.0;
      c[j] = 0.0;
      b[j] = j;
    }

#pragma omp target enter data map(to:a[0:num_elements])
#pragma omp target enter data map(to:b[0:num_elements])
#pragma omp target enter data map(to:c[0:num_elements])

  daxpy( a, b, scalar, num_elements );

  daxpy( c, a, scalar, num_elements );

#pragma omp target update from(c[0:num_elements])

  // error checking
  for (size_t j=0; j<num_elements; j++) {
      if( fabs(c[j] - (double)j*scalar*scalar) > 0.000001  ) {
	  num_errors++;
	}
    }

#pragma omp target exit data map(release:c[0:num_elements])
#pragma omp target exit data map(release:a[0:num_elements])
#pragma omp target exit data map(release:b[0:num_elements])

  free(a);
  free(b);
  free(c);

  if(num_errors == 0) printf( "Success!\n" );

  assert(num_errors == 0);

  return 0;
}
